src/cpp/src/storage/checkpointer.cpp

```diff
@@ -15,10 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
     config_ = config;
 }
 
-void Checkpointer::save(string checkpoint_dir, CheckpointMeta checkpoint_meta) {
+void Checkpointer::create_checkpoint(string checkpoint_dir, CheckpointMeta checkpoint_meta, int epochs) {
+    string tmp_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "_tmp/";
+    std::filesystem::create_directory(tmp_checkpoint_dir);
 
-    if (checkpoint_meta.has_model) {
+    std::string new_embeddings_file = tmp_checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string new_embeddings_state_file = tmp_checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    std::string embeddings_file = checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string embeddings_state_file = checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    if (std::filesystem::exists(embeddings_file)) {
+        if (this->config_->save_encoded) std::filesystem::copy(embeddings_file, new_embeddings_file);
+        if (this->config_->save_state) std::filesystem::copy(embeddings_state_file, new_embeddings_state_file);
+    }
 
+    this->save(tmp_checkpoint_dir, checkpoint_meta);
+
+    string final_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "/";
+    std::filesystem::rename(tmp_checkpoint_dir, final_checkpoint_dir);
+
+    int delete_checkpoint = epochs - this->config_->save_prev_num * this->config_->interval;
+    if (delete_checkpoint > 0) {
+        std::string delete_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(delete_checkpoint) + "/";
+        std::filesystem::remove_all(delete_checkpoint_dir);
+    }
+}
+
+void Checkpointer::save(string checkpoint_dir, CheckpointMeta checkpoint_meta) {
+    if (checkpoint_meta.has_model) {
```

Comment:
The CheckpointMeta struct and corresponding file could be cleaned up. Can you 1) delete the unused options in CheckpointMeta and 2) have the metadata file include the keys so it is more interpretable. 

I believe the current format just has the values and no keys. So the new version should be something like

```
has_state: 1
link_prediction: 1
...
```

However adding the keys to the checkpoint metadata file may make the parsing code significantly more complicated (C++ isn't great at this stuff), so if this is the case you can skip that part and just delete the unused options.